devicetypes/sky-nie/evalogik-door-window-sensor.src/evalogik-door-window-sensor.groovy

```diff
@@ -0,0 +1,675 @@
+}
+
+private getAssociationGroupSettingParam() {
+	return getParam(19, "Association Group 2 Setting", 1, 1, associationGroupSettingOptions)
+}
+
+private getParam(num, name, size, defaultVal, options=null, range=null) {
+	def val = safeToInt((settings ? settings["configParam${num}"] : null), defaultVal)
+
+	def map = [num: num, name: name, size: size, value: val]
+	if (options) {
+		map.valueName = options?.find { k, v -> "${k}" == "${val}" }?.value
+		map.options = setDefaultOption(options, defaultVal)
+	}
+	if (range) {
+		map.range = range
+	}
+
+	return map
+}
+
+private static setDefaultOption(options, defaultVal) {
+	return options?.collectEntries { k, v ->
+		if ("${k}" == "${defaultVal}") {
+			v = "${v} [DEFAULT]"
+		}
+		["$k": "$v"]
+	}
+}
+
+// Setting Options
+private static getSwitchTemperatureUnitOptions() {
+	return [
+		"0":"Celsius",
+		"1":"Fahrenheit"
+	]
+}
+
+private static getAssociationGroupSettingOptions() {
+	return [
+		"0":"Disable completely",
+		"1":"Send Basic SET 0xFF when Magnet is away,and send Basic SET 0x00 when Magnet is near.",
+		"2":"Send Basic SET 0x00 when Magnet is away,and send Basic SET 0xFF when Magnet is near",
+		"3":"Only send Basic SET 0xFF when Magnet is away",
+		"4":"Only send Basic SET 0x00 when Magnet is near",
+		"5":"Only send Basic SET 0x00 when Magnet is away",
+		"6":"Only send Basic SET 0xFF when Magnet is near"
+	]
+}
+
+private static getSensorModeWhenCloseOptions() {
+	return [
+		"0":"door/window closed",
+		"1":"door/window opened"
+	]
+}
+
+private static getNotificationAndAssociationGroupControlOptions(int groupId){
+	return [
+		"0":"disable notification and association group basic set",
+		"1":"only notification report to lifeline group",
+		"2":"only basic set on to association group ${groupId}",
+		"3":"notification to lifeline and basic set on to association group ${groupId}",
+		"4":"only basic set off to association group ${groupId}",
+		"5":"notification to lifeline and basic off to association group ${groupId}",
+		"6":"basic set on and off to association group ${groupId}",
+		"7":"notification to lifeline and basic set on and off to association group ${groupId}"
+	]
+}
+
+def sensorValueEvent(value) {
+	if (value) {
+		createEvent(name: "contact", value: "open", descriptionText: "$device.displayName is open")
+	} else {
+		createEvent(name: "contact", value: "closed", descriptionText: "$device.displayName is closed")
+	}
+}
+
+private static safeToInt(val, defaultVal=0) {
+	return "${val}"?.isInteger() ? "${val}".toInteger() : defaultVal
+}
+
+private convertToLocalTimeString(dt) {
+	def timeZoneId = location?.timeZone?.ID
+	if (timeZoneId) {
+		return dt.format("MM/dd/yyyy hh:mm:ss a", TimeZone.getTimeZone(timeZoneId))
+	} else {
+		return "$dt"
+	}
+}
+
+private static isDuplicateCommand(lastExecuted, allowedMil) {
+	!lastExecuted ? false : (lastExecuted + allowedMil > new Date().time)
+}
+
+private logDebug(msg) {
```

Comment:
Are those functions really needed: logDebug, logTrace ?

They seem to be redundant:
```
logDebug "installed()..."
log.debug "installed()..."
```

Reply:
I have modified them and updated this branch.